### ISSUE TITLE
NDPI_LIB_COMPILATION needs to be defined in order to build an application with nDPI

### ISSUE THREAD

**Issue by pavlosantoniou** [ntop/nDPI#605]
This issue has been observed on the 2.4-stable release of nDPI.

The nDPI release builds successfully, but when the nDPI headers are included in a project, the project's build fails due to compiler errors such as:

> error: invalid use of incomplete type 'struct ndpi_detection_module_struct'
> error: forward declaration of 'struct ndpi_detection_module_struct'

The solution is to define the pre-processor macro NDPI_LIB_COMPILATION when building the project that includes the nDPI headers. Should this be like this?

**Comment by lucaderi** [comment]
Can you please explain how to reproduce the bug? For instance ntopng does not show this issue during compilation, so I think we're not at the same page and I need to understand how to reproduce it. Thanks

**Comment by dsokoloski** [comment]
This isn't a bug -- it's a feature! ;)

The intention (my best guess), is that `struct ndpi_detection_module_struct` is intended to be opaque.  That is, those that link to this library are really not supposed to be directly accessing members of `struct ndpi_detection_module_struct`.  There are many API functions that are provided to do most things.

If you however require complete access to this structure, then only having a forward-declaration is insufficient and will fail to compile as soon as you reference a structure member.  However, simply defining `NDPI_LIB_COMPILATION` before including the API (`ndpi_api.h`) allows complete access -- which is what I do in my application.

The issue as reported is easy to reproduce using the included ndpiReader example:

- Edit: `./example/ndpi_util.c` and remove the conditional `#ifdef HAVE_CONFIG_H` surrounding the `#include "ndpi_config.h"` directive.  By the way, since the destruction of the autoconf implementation, this header is no longer being included in `./example/ndpiReader.c`, nor `./example/ndpi_util.c`.  Not sure if they need it.  This is why nobody noticed this bug in ndpiReader I assume.
- Re-configure the source but enable debug messages: `./configure --enable-debug-messages`
- Recompile: `make clean all`

This will fail because there is a section of code in `./example/ndpi_util.c` that tries to access the `debug_bitmask` member of `struct ndpi_detection_module_struct`, which can not be properly resolved without  `NDPI_LIB_COMPILATION` being defined because `struct ndpi_detection_module_struct` is merely a forward-declaration.

In short (I know... too late): `-DNDPI_LIB_COMPILATION`

**Comment by pavlosantoniou** [comment]
@dsokoloski I agree with the opaque struct rationale.

The confusion will most probably arise when (as in my case):
- An application is being compiled successfully with an older nDPI version
- The nDPI version is bumped, and the application is compiled against it
- Now the error appears, there seems to be a problem with something as basic as `ndpi_detection_module_struct `
- The user has to dig into the nDPI code to find out about the macro

I believe that the need for defining `NDPI_LIB_COMPILATION` could be a note in a 'how to compile with nDPI' section in the readme page.

**Comment by IvanNardi** [comment]
Compilation with `--enable-debug-message` has been fixed in 56d87186.
No know issues on compiling current nDPI (and example applications). Closing
